### PR TITLE
BTree Index Scanner with Partial Loading

### PR DESCRIPTION
**Pull request by linhong-intel**
## What changes were proposed in this pull request?

BTree Index Scanner
Based on #346

## How was this patch tested?

Unit test

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexFileReader.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    (fs.open(file), fs.getFileStatus(file).getLen)
+  }
+
+  private val (footerLength, rowIdListLength) = {
+    val sectionLengthIndex = fileLength - FOOTER_LENGTH_SIZE - ROW_ID_LIST_LENGTH_SIZE
+    val sectionLengthBuffer = new Array[Byte](FOOTER_LENGTH_SIZE + ROW_ID_LIST_LENGTH_SIZE)
+    reader.readFully(sectionLengthIndex, sectionLengthBuffer)
+    val rowIdListSize = getIntFromBuffer(sectionLengthBuffer, 0)
+    val footerSize = getIntFromBuffer(sectionLengthBuffer, ROW_ID_LIST_LENGTH_SIZE)
+    (footerSize, rowIdListSize)
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexFileReader.scala]: Yes. But that size is not decided. Let's keep the easiest and safest way and make improvement step by step.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+      else e = m - 1
+    }
+    (m, found)
+  }
+
+  private lazy val partialOrdering = GenerateOrdering.create(StructType(schema.dropRight(1)))
+  private def rowOrdering(x: InternalRow, y: InternalRow, isStart: Boolean): Int = {
+    if (x.numFields == y.numFields) {
+      ordering.compare(x, y)
+    } else if (x.numFields < y.numFields) {
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala]: `select * from t where a = 1 and b > 0`
Then
`interval.start = (a = 1, b = 0)`
`interval.end = (a = 1, b = DUMMY_END) and DUMMY_END is actually nothing`
And assume there is a row in the table `a = 1, b = 2`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+  private var internalIterator: Iterator[Int] = _
+
+  private var footer: BTreeFooter = _
+  private var rowIdList: BTreeRowIdList = _
+
+  private var reader: BTreeIndexFileReader = _
+
+  @transient private lazy val ordering = GenerateOrdering.create(schema)
+
+  def initialize(path: Path, intervalArray: ArrayBuffer[RangeInterval]): Unit = {
```
[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala]: Yes, but we have no bandwidth on it until we finish current job.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (yueawang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala]: These x & y are intervals? I thought they are rows from database.

[the remark this answers is not in the document]
Comment (linhong-intel) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BTreeIndexRecordReader.scala]: x is a interval, y is a row. OR x is a row, y is a interval

**Pull request merged**